### PR TITLE
T38 json input controllers

### PR DESCRIPTION
**Pull request by Gregam3**

### CODE REVIEW

File: src/main/scala/entities/module/Module.scala [hunk 1]
```diff
@@ -1,10 +1,21 @@
...
 
 import scala.collection.mutable
 import scala.collection.mutable.ListBuffer
 
+object Module {
+  def apply(json:Map[String, Any]): Module = {
+    if(json("moduleId") == null) throw new InvalidJsonException("Missing Id")
+
+    new Module(extractInt(json("moduleId")), extractString(json("moduleCode")), extractString(json("moduleName")),
+      extractString(json("moduleDescription")), null, ListBuffer(), mutable.Set())
```
[the remark this answers is not in the document]
Comment (jamesfernando94) [reply src/main/scala/entities/module/Module.scala]: Could this be an empty list?

File: conf/routes [hunk 2]
```diff
@@ -6,12 +6,16 @@
 # An example controller showing a sample home page
 GET     /                           controllers.HomeController.index
 
-GET     /get-generated-schedule                           controllers.ScheduleController.generateScheduleResponse
+GET     /get-generated-schedule                 controllers.ScheduleController.generateScheduleResponse
+GET     /insert-student                         controllers.InsertController.insertStudent
```
Comment (jamesfernando94) [inline conf/routes]: Would it not be more correct to have a route which takes the form `/student/insert` instead?

File: src/main/scala/entities/people/Student.scala [hunk 3]
```diff
@@ -4,17 +4,25 @@ import java.util.regex.Pattern
...
+import services.JsonUtils._
 import scala.collection.mutable
 import scala.collection.mutable.ListBuffer
 import scala.io.Source
 import scala.util.Random
 
 
 object Student extends Generator[Student] {
+  def apply(json:Map[String, Any]): Student = {
+    if(json("studentId") == null) throw new InvalidJsonException("Missing Id")
```
Comment (jamesfernando94) [inline src/main/scala/entities/people/Student.scala]: Surely if we don't provide an ID when adding a student we want one to be automatically generated?

File: conf/routes (hunk 2 above)
Comment (jamesfernando94) [reply conf/routes]: Also could we have it as a post so I can send the student data in the body

File: src/main/scala/entities/people/Student.scala (hunk 3 above)
Comment (Gregam3) [reply src/main/scala/entities/people/Student.scala]: Yeah okay I'll change this

File: src/main/scala/db/Dao.scala [hunk 4]
```diff
@@ -4,26 +4,53 @@ import java.io.{ByteArrayInputStream, ByteArrayOutputStream, ObjectInputStream,
...
+    override def handle(rs: ResultSet): Array[AnyRef] = if(rs.next) Array(Student(rs)) else Array()
+  }
+
+  val moduleHandler = new ResultSetHandler[Array[AnyRef]]() {
+    @throws[SQLException]
+    override def handle(rs: ResultSet): Array[AnyRef] = if(rs.next) Array(Module(rs)) else Array()
+  }
+
+  //Even though these are set in environment variables there is currently a bug in intellij that sets env variables as properties instead in play
+  private[this] val conn =  DriverManager.getConnection(sys.props("DB_URL"), sys.props("DB_USER"), sys.props("DB_PASSWORD"))
```
[the remark this answers is not in the document]
Comment (Gregam3) [reply src/main/scala/db/Dao.scala]: Yeah good find dumbass

**Comment by jamesfernando94** [comment]
I Get a Compilation Error when running it this is after I locally merged it with master
file

```
Information:23/04/2019 21:12 - Compilation completed with 3 errors and 0 warnings in 3 s 665 ms
C:\GIT\Tempus\src\main\scala\db\TimeTableDao.scala
Error:(17, 21) overloaded method value query with alternatives:
  [T](x$1: String, x$2: org.apache.commons.dbutils.ResultSetHandler[T], x$3: Object*)T <and>
  [T](x$1: java.sql.Connection, x$2: String, x$3: org.apache.commons.dbutils.ResultSetHandler[T], x$4: Object*)T <and>
  [T](x$1: java.sql.Connection, x$2: String, x$3: Array[Object], x$4: org.apache.commons.dbutils.ResultSetHandler[T])T <and>
  [T](x$1: java.sql.Connection, x$2: String, x$3: Any, x$4: org.apache.commons.dbutils.ResultSetHandler[T])T
 cannot be applied to (java.sql.Connection, String, org.apache.commons.dbutils.ResultSetHandler[Array[AnyRef]], Int)
    deserialize(run.query(conn, "SELECT * FROM TIMETABLES WHERE ID=?", handler, id).head.asInstanceOf[Array[Byte]])
C:\GIT\Tempus\src\main\scala\entities\module\Module.scala
Error:(17, 5) overloaded method constructor Module with alternatives:
  (name: String)entities.module.Module <and>
  ()entities.module.Module <and>
  (moduleId: Int,moduleCode: String,moduleName: String,moduleDescription: String,school: entities.School,terms: scala.collection.mutable.ListBuffer[Int],requiredSessions: scala.collection.mutable.Set[entities.module.RequiredSession],sessionStructure: scala.collection.mutable.ListBuffer[entities.module.ModuleSessionStructure])entities.module.Module
 cannot be applied to (Int, String, String, String, Null, scala.collection.mutable.ListBuffer[Nothing], scala.collection.mutable.Set[Nothing])
    new Module(moduleId, extractString(json("moduleCode")), extractString(json("moduleName")),
Error:(21, 38) overloaded method constructor Module with alternatives:
  (name: String)entities.module.Module <and>
  ()entities.module.Module <and>
  (moduleId: Int,moduleCode: String,moduleName: String,moduleDescription: String,school: entities.School,terms: scala.collection.mutable.ListBuffer[Int],requiredSessions: scala.collection.mutable.Set[entities.module.RequiredSession],sessionStructure: scala.collection.mutable.ListBuffer[entities.module.ModuleSessionStructure])entities.module.Module
 cannot be applied to (Int, String, String, String, Null, Null, Null)
  def apply(qr: ResultSet): Module = new Module(qr.getInt(1), qr.getString(2), qr.getString(3), qr.getString(4), null, null, null)
```

**Comment by jamesfernando94** [comment]
Otherwise, I didn't find any errors while reading through the code

**Pull request merged**